feature: dynamic prefers color scheme

gordonwoodhull opened the pull request:
This is a case of "it was easier to fix this than to document it the way it is".

We can watch the media query `prefers-color-scheme` in exactly the same way we watch the visual toggle.

<strike>This does not introduce any new FOUC. However we still have a FOUC on some browsers when toggling dark mode dynamically #12309.</strike> This turns out to be a `quarto preview` issue.

I am hesitant to merge this, but it would make documenting `respect-user-color-scheme` ***much*** easier if we don't have to say it's only on page load and not dynamic.

First let's see if tests pass. I have added a bunch of them in this PR.

gordonwoodhull committed:
watch prefers-color-scheme

it's the same as the visual toggle except we don't set local storage

gordonwoodhull committed:
remove extra class setting

we set quarto-light or quarto-dark at the end of this script,
with the correct setting. so there is no need to set it here.

gordonwoodhull committed:
set visual toggle once page is loaded

also, exhaustively test visual toggle

gordonwoodhull commented:
This has an annoying interaction with the visual toggle. Right now they both "win" in the sense that changing the OS setting always changes the appearance when `respect-user-color-scheme` is set, but the local storage setting from the toggle is still saved and takes precedence on reload.

It's like a 1-bit CRDT problem.

I see three ways to make them consistent:
1. the OS setting should have no effect when there is a local storage setting
2. option 1, but the local storage setting should get removed when it matches the OS setting
3. the local storage setting should get removed when the OS setting changes

1 is simplest, but you can't ever get out once you've clicked the toggle, without going into dev tools and deleting the local storage key.

I think 2 is my favorite but it's complicated both to implement and to document.

3 is simple but kind of weird and hard to explain.

cscheid commented:
Huh, yeah, this is all a bit weird. I wonder what other SSGs that have dark-light mode do in this case...

gordonwoodhull commented:
According to [The Complete Guide to the Dark Mode Toggle](https://ryanfeigenbaum.com/dark-mode/) 

> If the user has set an OS-level preference for a color scheme and has **not** set a preference on our site during an earlier visit, then we should respect the OS-level preference and render the site accordingly in light or dark mode.

This is option 1 and the simplest. I think I'll just do this, and we can revisit if we want to be even more respectful of the user's OS setting in the future.

gordonwoodhull commented:
From a code review standpoint, this PR adds two more entries to `window` (in order to update the visual toggle after the page has loaded). Is this okay?

gordonwoodhull committed:
do not respect user color scheme if user has set a preference

also, the localAlternateSentinel is used dynamically
so it needs to be updated when we detect an OS change
so that the toggle works on the first click

gordonwoodhull commented:
I'll go ahead merge this, and ask my code review question on Thursday.

Glad to go back and clean this up later.
